### PR TITLE
Fix serialization verification problem with Akka.IO messages

### LINKED ISSUE #4936: During CoordinatedShutdown an exception 'Hyperion.ValueSerializers.UnsupportedTypeException: No coercion operator is defined between types 'Interop+Sys+SocketEvent*' and 'System.Object'.' is thrown
I got a similar issue as mentioned in #4860. I tested this with the nightly version of 1.4.19 as there #4860 is fixed.

```
<PackageReference Include="Akka" Version="1.4.19-beta637539553808642405" />
<PackageReference Include="Akka.Cluster" Version="1.4.19-beta637539553808642405" />
<PackageReference Include="Akka.Cluster.Sharding" Version="1.4.19-beta637539553808642405" />
<PackageReference Include="Akka.Cluster.Tools" Version="1.4.19-beta637539553808642405" />
<PackageReference Include="Akka.DependencyInjection" Version="1.4.19-beta637539553808642405" />
<PackageReference Include="Akka.Persistence.Query.Sql" Version="1.4.19-beta637539553808642405" />
<PackageReference Include="Akka.Persistence.SqlServer" Version="1.4.16" />
<PackageReference Include="Akka.Serialization.Hyperion" Version="1.4.19-beta637539553808642405" />
```

But am still getting this error on shutdown of my ActorSystem. @Aaronontheweb mentioned that it was related to #4860 but it looks not related.

The error which is thrown is:
```
[ERROR][04/14/2021 19:38:05][Thread 0082][[akka://blueprint-service/system/IO-TCP#1876155147]] Swallowing exception during message send
Cause: Hyperion.ValueSerializers.UnsupportedTypeException: No coercion operator is defined between types 'Interop+Sys+SocketEvent*' and 'System.Object'.
   at Hyperion.ValueSerializers.UnsupportedTypeSerializer.WriteManifest(Stream stream, SerializerSession session)
   at Hyperion.Extensions.StreamEx.WriteObject(Stream stream, Object value, Type valueType, ValueSerializer valueSerializer, Boolean preserveObjectReferences, SerializerSession session)
   at Hyperion.Extensions.StreamEx.WriteObject(Stream stream, Object value, Type valueType, ValueSerializer valueSerializer, Boolean preserveObjectReferences, SerializerSession session)
   at lambda_method(Closure , Stream , Object , SerializerSession )
   at Hyperion.Extensions.StreamEx.WriteObject(Stream stream, Object value, Type valueType, ValueSerializer valueSerializer, Boolean preserveObjectReferences, SerializerSession session)
   at lambda_method(Closure , Stream , Object , SerializerSession )
   at Hyperion.Extensions.StreamEx.WriteObject(Stream stream, Object value, Type valueType, ValueSerializer valueSerializer, Boolean preserveObjectReferences, SerializerSession session)
   at lambda_method(Closure , Stream , Object , SerializerSession )
   at Hyperion.Extensions.StreamEx.WriteObject(Stream stream, Object value, Type valueType, ValueSerializer valueSerializer, Boolean preserveObjectReferences, SerializerSession session)
   at lambda_method(Closure , Stream , Object , SerializerSession )
   at Hyperion.ValueSerializers.ObjectSerializer.<.ctor>b__9_1(Stream stream, Object o, SerializerSession session)
   at Akka.Serialization.HyperionSerializer.ToBinary(Object obj)
   at Akka.Actor.ActorCell.SerializeAndDeserializePayload(Object obj)
   at Akka.Actor.ActorCell.SerializeAndDeserialize(Envelope envelope)
   at Akka.Actor.ActorCell.SendMessage(Envelope message)
```

This is thrown when a CoordinatedShutdown is initiated from the ActorSystem. I can also reproduce this in 1.4.18 (not sure about 1.4.17 but in our stable branch we use 1.4.16 and there it doesn't happen).

### PR DESCRIPTION
**Pull request by Arkatufus**
Closes #4936

### CODE REVIEW

File: src/core/Akka.Tests/IO/TcpIntegrationSpec.cs [hunk 1]
```diff
@@ -45,6 +45,8 @@ class AckWithValue : Tcp.Event
         
         public TcpIntegrationSpec(ITestOutputHelper output)
             : base($@"akka.loglevel = DEBUG
+                     akka.actor.serialize-creators = on
+                     akka.actor.serialize-messages = on
```
Comment (Arkatufus) [inline src/core/Akka.Tests/IO/TcpIntegrationSpec.cs:49]: Leaving this here so we can test similar problems in the future

File: src/core/Akka.Tests/IO/TcpListenerSpec.cs [hunk 2]
```diff
@@ -19,12 +19,14 @@ namespace Akka.Tests.IO
     public class TcpListenerSpec : AkkaSpec
     {
         public TcpListenerSpec()
-            : base(@"akka.io.tcp.register-timeout = 500ms
+            : base(@"
+                     akka.actor.serialize-creators = on
+                     akka.actor.serialize-messages = on
```
Comment (Arkatufus) [inline src/core/Akka.Tests/IO/TcpListenerSpec.cs:24]: Leaving this here so we can test similar problems in the future

File: src/core/Akka.Tests/IO/UdpConnectedIntegrationSpec.cs [hunk 3]
```diff
@@ -24,6 +24,8 @@ public class UdpConnectedIntegrationSpec : AkkaSpec
 
         public UdpConnectedIntegrationSpec(ITestOutputHelper output)
             : base(@"
+                    akka.actor.serialize-creators = on
+                    akka.actor.serialize-messages = on
```
Comment (Arkatufus) [inline src/core/Akka.Tests/IO/UdpConnectedIntegrationSpec.cs:28]: Leaving this here so we can test similar problems in the future

File: src/core/Akka.Tests/IO/UdpIntegrationSpec.cs [hunk 4]
```diff
@@ -27,6 +27,8 @@ public class UdpIntegrationSpec : AkkaSpec
 
         public UdpIntegrationSpec(ITestOutputHelper output)
             : base(@"
+                    akka.actor.serialize-creators = on
+                    akka.actor.serialize-messages = on
```
Comment (Arkatufus) [inline src/core/Akka.Tests/IO/UdpIntegrationSpec.cs:31]: Leaving this here so we can test similar problems in the future

File: src/core/Akka.Tests/IO/TcpIntegrationSpec.cs [hunk 5]
```diff
@@ -190,7 +192,7 @@ public void The_TCP_transport_implementation_should_properly_support_connecting_
             var targetAddress = new DnsEndPoint("localhost", boundMsg.LocalAddress.AsInstanceOf<IPEndPoint>().Port);
             var clientHandler = CreateTestProbe();
             Sys.Tcp().Tell(new Tcp.Connect(targetAddress), clientHandler);
-            clientHandler.ExpectMsg<Tcp.Connected>(TimeSpan.FromMinutes(10));
```
Comment (Arkatufus) [inline src/core/Akka.Tests/IO/TcpIntegrationSpec.cs:193]: A socket shouldn't take 10 minutes to connect

File: src/core/Akka.Tests/IO/UdpListenerSpec.cs [hunk 6]
```diff
@@ -13,17 +13,21 @@
...
 namespace Akka.Tests.IO
 {
     public class UdpListenerSpec : AkkaSpec
     {
-        public UdpListenerSpec()
-            : base(@"akka.io.udp.max-channels = unlimited
+        public UdpListenerSpec(ITestOutputHelper output)
+            : base(@"
+                    akka.actor.serialize-creators = on
+                    akka.actor.serialize-messages = on
```
Comment (Arkatufus) [inline src/core/Akka.Tests/IO/UdpListenerSpec.cs:26]: Leaving this here so we can test similar problems in the future

File: src/core/Akka/Actor/ActorCell.Children.cs [hunk 7]
```diff
@@ -465,17 +468,24 @@ private IInternalActorRef MakeChild(Props props, string name, bool async, bool s
...
-                                            nameof(name));
+                                        $"Pre-creation serialization check failed at [${_self.Path}/{name}]",
+                                        nameof(name));
                             }
                         }
                     }
                 }
+                catch (Exception e)
+                {
+                    throw new SerializationException(
```
Comment (Arkatufus) [inline src/core/Akka/Actor/ActorCell.Children.cs:485]: Catch and re-throw to add a more meaningful exception message to the error

File: src/core/Akka/Actor/ActorCell.cs [hunk 8]
```diff
@@ -540,6 +544,10 @@ private object SerializeAndDeserializePayload(object obj)
                 var manifest = Serialization.Serialization.ManifestFor(serializer, obj);
                 return _systemImpl.Serialization.Deserialize(bytes, serializer.Identifier, manifest);
             }
+            catch (Exception e)
+            {
+                throw new SerializationException($"Failed to serialize and deserialize payload object {obj.GetType()}", e);
```
Comment (Arkatufus) [inline src/core/Akka/Actor/ActorCell.cs:549]: Catch and re-throw to add a more meaningful exception message to the error

File: src/core/Akka/Actor/ActorCell.cs [hunk 9]
```diff
@@ -518,6 +519,9 @@ private Envelope SerializeAndDeserialize(Envelope envelope)
             if (unwrapped is INoSerializationVerificationNeeded)
                 return envelope;
 
+            if(unwrapped.GetType().Namespace?.StartsWith("System.Net.Sockets") ?? false)
```
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/core/Akka/Actor/ActorCell.cs:522]: I'd prefer it if we just wrapped the `SocketAsyncEventArgs` inside a `struct` of some kind wrapped with `INoSerializationVerificationNeeded`

File: src/core/Akka/Actor/ActorCell.cs [hunk 10]
```diff
@@ -544,10 +550,6 @@ private object SerializeAndDeserializePayload(object obj)
                 var manifest = Serialization.Serialization.ManifestFor(serializer, obj);
                 return _systemImpl.Serialization.Deserialize(bytes, serializer.Identifier, manifest);
             }
-            catch (Exception e)
```
Comment (Arkatufus) [inline src/core/Akka/Actor/ActorCell.cs]: Move the exception handling to the caller function so we can also capture the envelope

File: src/core/Akka/IO/TcpListener.cs [hunk 11]
```diff
@@ -71,9 +71,9 @@ private IEnumerable<SocketAsyncEventArgs> Accept(int limit)
             {
                 var self = Self;
                 var saea = new SocketAsyncEventArgs();
-                saea.Completed += (s, e) => self.Tell(e);
+                saea.Completed += (s, e) => self.Tell(new SocketEvent(e));
```
Comment (Arkatufus) [inline src/core/Akka/IO/TcpListener.cs:74]: Wrap naked SocketAsyncEventArgs in a struct

File: src/core/Akka/IO/TcpListener.cs [hunk 12]
```diff
@@ -130,5 +130,15 @@ protected override void PostStop()
                 _log.Debug("Error closing ServerSocketChannel: {0}", e);
             }
         }
+
+        private struct SocketEvent : INoSerializationVerificationNeeded
```
Comment (Arkatufus) [inline src/core/Akka/IO/TcpListener.cs:134]: SocketAsyncEventArgs wrapper

File: src/core/Akka/IO/TcpListener.cs (hunk 11 above)
Comment (Aaronontheweb) [reply src/core/Akka/IO/TcpListener.cs:74]: :+1:

File: src/core/Akka/Actor/ActorCell.Children.cs [hunk 13]
```diff
@@ -465,17 +468,24 @@ private IInternalActorRef MakeChild(Props props, string name, bool async, bool s
...
+                        $"Failed to serialize and deserialize actor props argument of type {propArgument?.GetType()}",
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/ActorCell.Children.cs:486]: It's be a good idea to capture the actor type here too

File: src/core/Akka/Actor/ActorCell.cs [hunk 14]
```diff
@@ -518,7 +519,16 @@ private Envelope SerializeAndDeserialize(Envelope envelope)
...
 
-            var deserializedMsg = SerializeAndDeserializePayload(unwrapped);
+            object deserializedMsg;
+            try
+            {
+                deserializedMsg = SerializeAndDeserializePayload(unwrapped);
+            }
+            catch (Exception e)
+            {
+                throw new SerializationException($"Failed to serialize and deserialize payload object [{unwrapped.GetType()}]. Envelope: [{envelope}]", e);
```
Comment (Aaronontheweb) [inline src/core/Akka/Actor/ActorCell.cs:529]: Record the actor type too

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Should clean up the serialization exceptions to include actor type in addition to message / argument type

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**